TypeSystem/flow-analysis/promotion_via_assignment_A07_t02.dart

```diff
@@ -2,17 +2,85 @@
+///       `written <: T <: provisionalType`.
+///     - If `p3` contains exactly one type `T` that is a subtype of all the
+///       others, then `promoted` is `[...promotionChain, T]`. Writing a value
+///       whose static type is a subtype of a type of interest promotes to that
+///       type of interest, provided there is a single "best" type of interest
+///       available to promote to.
+///       - Since `T <: provisionalType <: declared`, and all types `U` in
+///         `promotionChain` satisfy `provisionalType <: U`, it follows that all
+///         types `U` in `promotionChain` satisfy `T <: U`. Therefore
+///         `newPromotionChain` satisfies the definition of a promotion chain,
+///         and is valid for declared type `declared`.
+///     - Otherwise, `newPromotionChain` is `promotionChain`. If there is no
+///       single "best" type of interest to promote to, then no type of interest
+///       promotion is done.
 ///
-/// @description Checks that a variable is not promotable if `T` is not a
-/// subtype of `S`. Test extension types.
+/// `assign(x, E, M)` where `x` is a local variable, `E` is an
+/// expression of inferred type `T` (which must be a subtype of `x`'s declared
+/// type), and `M = FlowModel(r, VI)` is the flow model for `E` is defined to be
+/// `FlowModel(r, VI[x -> VM])` where:
+/// - `VI(x) = VariableModel(declared, promoted, tested, assigned, unassigned, captured)`
+/// - If `captured` is true then:
+///   - `VM = VariableModel(declared, promotionChain, tested, true, false, captured)`.
+/// - Otherwise:
+///   - Let `written = T`.
+///   - Let `promotionChain' = demote(promotionChain, written)`.
+///   - Let `promotionChain'' = toi_promote(declared, promotionChain', tested, written)`.
+///     - The preconditions for toi_promote are satisfied because:
+///       - `demote` deletes any elements from `promotionChain` that do not
+///         satisfy `written <: T`, therefore every element of `promotionChain'`
+///         satisfies `written <: T`.
+///       - `written = T` and `T` is a subtype of `x`'s declared type, therefore
+///         `written <: declared`.
+///   - Then `VM = VariableModel(declared, promotionChain'', tested, true, false, captured)`.
+/// ...
+/// Local-variable assignment: If `N` is an expression of the form `x = E1`
+/// where `x` is a local variable, then:
+/// - Let `before(E1) = before(N)`.
+/// - Let `E1'` be the result of applying type coercion to `E1`, to coerce it to
+///   the declared type of `x`.
+/// - Let `after(N) = assign(x, E1', after(E1))`.
+///   - Since type coercion to type `T` produces an expression whose static type
+///     is a subtype of `T`, the precondition of assign is satisfied, namely
+///     that the static type of `E1'` must be a subtype of `x`'s declared type.
+///
+/// @description Checks that a variable is not promotable from the type `S` to
+/// the type `T` if `T` is not a subtype of `S`. Test extension types.
```

Comment:
We can't do exactly the same thing here because the extension types aren't subtypes of the declared type. We could adjust the extension types such that it isn't an error, or we could drop the test if it turns out to be to similar to the previous one.